src/apis/orderbookapi.py

```diff
@@ -0,0 +1,173 @@
+from src.models import Trade, OrderExecution
+from src.constants import (
+    header,
+    REQUEST_TIMEOUT,
+    SUCCESS_CODE,
+    FAIL_CODE,
+)
+
+
+class OrderbookAPI:
+    """
+    Class for fetching data from a Web3 API.
+    """
+
+    def __init__(self):
+        self.logger = get_logger()
+
+    def get_solver_competition_data(self, tx_hash: str) -> Optional[dict[str, Any]]:
+        """
+        Get solver competition data from a transaction hash.
+        The returned dict follows the schema outlined here:
+        https://api.cow.fi/docs/#/default/get_api_v1_solver_competition_by_tx_hash__tx_hash_
+        """
+        prod_endpoint_url = (
+            "https://api.cow.fi/mainnet/api/v1/solver_competition"
+            f"/by_tx_hash/{tx_hash}"
+        )
+        barn_endpoint_url = (
+            "https://barn.api.cow.fi/mainnet/api/v1"
+            f"/solver_competition/by_tx_hash/{tx_hash}"
+        )
+        try:
+            json_competition_data = requests.get(
+                prod_endpoint_url,
+                headers=header,
+                timeout=REQUEST_TIMEOUT,
+            )
+            if json_competition_data.status_code == SUCCESS_CODE:
+                solver_competition_data = loads(json_competition_data.text)
+            elif json_competition_data.status_code == FAIL_CODE:
+                barn_competition_data = requests.get(
+                    barn_endpoint_url, headers=header, timeout=REQUEST_TIMEOUT
+                )
+                if barn_competition_data.status_code == SUCCESS_CODE:
+                    solver_competition_data = loads(barn_competition_data.text)
+            else:
+                return None
+        except requests.exceptions.ConnectionError as err:
+            self.logger.error(
+                f"Connection error while fetching competition data: {err}"
+            )
+            return None
+        return solver_competition_data
+
+    def get_solver_competition_data_list(
+        self,
+        tx_hash_list: list[str],
+    ) -> list[dict[str, Any]]:
+        """
+        This function uses a list of tx hashes to fetch and assemble competition data
+        for each of the tx hashes and returns it.
+        """
+        solver_competition_data_list = []
+        for tx_hash in tx_hash_list:
+            solver_competition_data = self.get_solver_competition_data(tx_hash)
+            if not solver_competition_data is None:
+                solver_competition_data_list.append(solver_competition_data)
+        return solver_competition_data_list
+
+    def get_endpoint_order_data(self, tx_hash: str) -> list[Any]:
+        """
+        Get all orders in a transaction from the transaction hash.
+        """
+        prod_endpoint_url = (
+            "https://api.cow.fi/mainnet/api/v1/transactions/" + tx_hash + "/orders"
+        )
+        barn_endpoint_url = (
+            "https://barn.api.cow.fi/mainnet/api/v1/transactions/" + tx_hash + "/orders"
+        )
```

Comment:
Nit. F-string. 

Also I am sending the base url being repeated here and in the previous function

Base urls can be constants defined at the top of the file